<img src="https://canada.unam.mx/wp-content/uploads/2018/12/FESAc.png" alt="Portada" style="float: left; margin-right: 50px" />

<h1>Licenciatura en Matematicas Aplicadas y Computación</h1>
<h1>Métodos Numéricos 2</h1>
<h2 style="color:#1a4996">Mtra. Teresa Carrillo Ramírez</h2>
<h2 style="color:#1a4996">Alfonso Flores Zenteno</h2>

___

# Regla del trapecio

Es un método sencillo que sirve para encontrar la integral de un función conocida en un intervalo determinado $[a,b]$. Por ejemplo tomemos la función:

$$f(x) = 10+12x - 15x^3$$

Y nos interesa encontrar la integral en el intervalo $[-0.5, 1.0]$ .

Podemos encontrar una aproximación a esta integral si calculamos el área del trapacio formado por los puntos $a, b, f(a), f(b)$, usando la fórmula para el área de un trapecio: 

$$A=\frac{B+b}{2}h$$

 Así tendremos una aproximación a la integral de esta manera :

Para calcular el área tenemos los siguientes datos:

| variable | sustitución | valor |
|---|---------|---------|
| B | f(1.0)  | 7.0     |
| b | f(-0.5) | 5.875   |
| h | (b-a)   | 1.5     | 
| A | $\frac{B+b}{2}h$ | 9.65625 |

Podemos ver que esta no es una aproximación exacta pero para encontrar una mejor solución se puede dividir el intervalo $[a, b]$ por la mitad y obtener dos trapecios diferentes. Al sumar el área de cada uno de estos tendremos un mejor acercamiento a la integral.

El área del primer trapecio será:
$$A=\frac{B+b}{2}h = \frac{f( \frac{a+b}{2} )+f(a)}{2}( \frac{a+b}{2} - a ) = \frac{f(0.25)+f(-0.5)}{2}0.75 = 6.990234$$

El área del segundo trapecio es:

$$A=\frac{B+b}{2}h = \frac{f( \frac{a+b}{2} )+f(b)}{2}( b - \frac{a+b}{2} ) = \frac{f(0.25)+f(1.0)}{2}0.75 = 7.412109$$

El resultado de sumar las áreas es: $14.402343$

Por lo que podemos apreciar una mejoria en el área cubierta por los trapecios y el valor numérico de la aproximación a la integral. La regla del trapecio nos permite seccionar el intervalo de nuestra función las veces que sean necesarias para crear más trapecios y mejorar la aproximación a la integral

Ahora que entendimos cómo funciona este método pasemos a aplicarlo en 4 pasos:

1. Modularizar las funciones y fórmulas a utilizar
2. Dividir el intervalo sobre el que se va a integrar
3. Calcular la aproximación a la integral
4. Comparar con distintos números de trapecios

## Modularizar funciones y fórmulas

En este paso definiremos la función $f(x) = 10+12x - 15x^3$ para evaluar puntos de manera rápida. También crearemos una función que calcule el área de un trapecio.

In [1]:
# La función polinómial que evaluaremos en este ejercicio se ve de la siguiente manera

def f1(x):
    return 10+(12*x)-(15*(x**3))

# Para evaluar un punto se llamaa la función de la siguiente manera
f1(1)

7

In [2]:
# La función que calcule el área de un trapecio a partir de sus medidas será la siguiente

def areaTrapecio(b1, b2, h):
    return ((b1+b2)/2)*h

# Podemos calcular el área a partir de las bases y la altura
areaTrapecio(5, 7.5, 4)

25.0

## Dividir el intervalo

Ahora crearemos el intervalo sobre el que se va a integrar, así como otras medidas útiles para realizar los cálculos más fáciles. Vamos a comparar el método con diferentes cantidades de diviciones sobre el intervalo, probaremos con 1, 2, 4 , 8, 16, 32, 64 y 128 trapecios.

In [3]:
# Ponemos el intervalo general en una tupla para que no sea modificable
intervalo = (-0.5, 1.0)

# Necesitamos conocer el largo total del intervalo para saber cómo dividirlo
largo = intervalo[1] - intervalo[0]

# Tendremos un arreglo con el número de diviciones o cortes del intervalo a probar
cortes = [1, 2, 4, 8, 16, 32, 64, 128]

In [4]:
# Usaremos la librería numpy para generar arreglos dando un valor inicial, uno final y
# el número de valores igualmente espaciados dentro del arreglo
import numpy as np

np.linspace(0, 10, 5)

array([ 0. ,  2.5,  5. ,  7.5, 10. ])

## Calcular la aproximación a la integral

Ahora que tenemos todos los elementos necesarios para calcular la integral es momento de usar los intervalos y las funciones para calcular el caso en el que se tienen 3 trapecios a partir del intervalo $[-0.5, 1.0]$

In [5]:
# Esta variable define cuántos trapecios pondremos debajo de la función
numTrapecios = 3

# Creamos un arreglo de números a partir del intervalo original en el que existan 3 pares de puntos consecutivos
subintervalo = np.linspace(intervalo[0], intervalo[1], 1+numTrapecios)

# Observamos los 3 pares de valores (-0.5, 0), (0, 0.5), (0.5, 1) que formarán las alturas de nuestros trapecios
print(subintervalo)

[-0.5  0.   0.5  1. ]


In [6]:
# Ya que los pares de puntos que tomaremos son equidistantes entre sí para todas las parejas, podemos calcular la
# altura de nuestros trapecios usando la primer pareja de puntos (-0.5, 0) y usarla para todos los casos
h = subintervalo[1]-subintervalo[0]

# Inicializamos la variable 'integral' en cero para acumular las sumas de los trapecios
integral = 0

# Creamos un ciclo que va de 0 a 3
for i in range(len(subintervalo)):
    
    # Ya que tomaremos el elemento actual en el índice 'i' y el siguiente 'i+1' debemos 
    # debemos solo considerar los elementos en los que 'i+1' no exceda el total de elementos dentro del arreglo
    if( i+1 < len(subintervalo) ):
        
        # En esta variable se acumula el cálculo del área del trapecio con los pares de intervalos válidos
        integral += areaTrapecio(f1(subintervalo[i]), f1(subintervalo[i+1]), h)

# Obtenemos la aproximación a la integral
print(integral)

15.28125


## Comparar con distintos números de trapecios

Para ilustrar los beneficios que tiene este método, calcularemos las aproximaciones con diferentes números de trapecios dentro del intervalo. Lo haremos iterando sobre el arreglo de 'cortes' que hicimos previamente.

In [7]:
integrales = [0]*len(cortes)

for n in range(len(cortes)):
    
    subintervalo = np.linspace(intervalo[0], intervalo[1], 1+cortes[n] )
    
    h = subintervalo[1]-subintervalo[0]
    
    resultado = 0
    
    for i in range(len(subintervalo)):
        
        if( i+1 < len(subintervalo) ):
            resultado += areaTrapecio(f1(subintervalo[i]), f1(subintervalo[i+1]), h)
    
    integrales[n] = resultado
    

print(integrales)

[9.65625, 14.40234375, 15.5888671875, 15.885498046875, 15.95965576171875, 15.978195190429688, 15.982830047607422, 15.983988761901855]


Llegamos a las siguientes aproximaciones dependiendo del número de trapecios:

| Número de trapecios | Aproximación |
|---------------------|--------------|
| 1 | 9.65625 |
| 2 | 14.402343 |
| 3 | 15.28125 |
| 4 | 15.5888671875 |
| 8 | 15.885498046875 |
| 16 | 15.95965576171875 |
| 32 | 15.978195190429688 |
| 64 | 15.982830047607422 |
| 128 | 15.983988761901855 |

Se aprecia que conforme aumentan el número de trapecios mejora la aproximación, desde la iteración con 3 trapecios se queda fija la parte entera en **15**, a partir de 16 trapecios se fija el primer número decimal y hasta el número 64 se fija el segundo valor decimal. Esto nos indica que hay una mejoria en la aproximación sin embargo se necesita de un tamaño de cortes grande para tener una aproximación más cercana al valor real, que en este caso es de $15.9844$